bftengine/src/communication/TlsTCPCommunication.cpp

```diff
@@ -956,54 +1231,73 @@ class TlsTCPCommunication::TlsTcpImpl {
+             string certRootFolder,
+             string cipherSuite,
+             UPDATE_CONNECTIVITY_FN statusCallback = nullptr) :
       _selfId(selfNodeNum),
       _listenPort(listenPort),
       _listenIp(listenIp),
       _bufferLength(bufferLength),
       _maxServerId(maxServerId),
-      _certRootFolder(certRootFolder) {
+      _certRootFolder(certRootFolder),
+      _logger(Logger::getLogger("concord.tls")),
+      _statusCallback{statusCallback},
+      _cipherSuite{cipherSuite} {
     //_service = new io_service();
     for (auto it = nodes.begin(); it != nodes.end(); it++) {
       _nodes.insert({it->first, it->second});
     }
 
-    if (_selfId < _maxServerId) {
-      tcp::endpoint ep(address::from_string(_listenIp), _listenPort);
-      _pAcceptor = boost::make_unique<tcp::acceptor>(_service, ep);
+    // all replicas are in listen mode
+    if (_selfId <= _maxServerId) {
+      // patch, we need to listen to all interfaces in order to support
+      // machines with internal/external IPs. Need to add "listen IP" to the BFT
+      // config file.
+      asio::ip::tcp::endpoint ep(
+          asio::ip::address::from_string(listenIp), _listenPort);
+      _pAcceptor = boost::make_unique<asio::ip::tcp::acceptor>(_service, ep);
       start_accept();
-    } else LOG_INFO("skipping listen for node: " << _selfId);
+    } else // clients don't listen
+    LOG_INFO(_logger, "skipping listen for node: " << _selfId);
 
+    // this node should connect only to nodes with lower ID
+    // and all nodes with higher ID will connect to this node
+    // we don't want that clients will connect to other clients
     for (auto it = _nodes.begin(); it != _nodes.end(); it++) {
       // connect only to nodes with ID higher than selfId
       // and all nodes with lower ID will connect to this node
-      if (it->first < _selfId && it->first < maxServerId) {
-        auto conn =
-            AsyncTlsConnection::create(
-                &_service,
-                std::bind(&TlsTcpImpl::on_async_connection_error,
-                          this,
-                          std::placeholders::_1),
-
-                std::bind(&TlsTcpImpl::on_hello_message,
-                          this,
-                          std::placeholders::_1,
-                          std::placeholders::_2),
-                _bufferLength,
-                it->first,
-                _selfId,
-                _certRootFolder,
-                ConnType::Outgoing);
-
-        _connections.insert(make_pair(it->first, conn));
-        string peerIp = std::get<0>(it->second);
-        uint16_t peerPort = std::get<1>(it->second);
-        conn->connect(peerIp, peerPort);
-        LOG_TRACE("connect called for node " << to_string(it->first));
+      if (it->first < _selfId && it->first <= maxServerId) {
+        create_outgoing_connection(it->first, it->second.ip, it->second.port);
       }
     }
   }
 
+  void create_outgoing_connection(
```

Comment:
reconnect() - is the function in AsyncTlsConnection - is called when connection to the peer is timed out and it should be re-connected (this is not the case when the LIVE connection has been broken, then this connection is disposed).
create_outgoing_connection() - is the function in the manager class, that creates outgoing connection object either on the beginning from the Ctor or after existing outgoing connection has been broken and disposed

Reply:
Ah, I got it. This will only be called after manager construction when _fOnError fires in the connection class. Then reconnect(), etc.. check that _closed is not set, or they just return, because the manager class is taking over with a new connection. I think part of my confusion was that `_closed` is a bit of a misnomer. The socket actually does get closed in `clean_for_reconnect()`. What the flag is really being used to check is that the socket is being disposed of.

I'd change `_close` to `_disposing` to make this clearer.

Thanks for the explanation. This does make sense to me now.